lib/Conversion/HandshakeToDC/HandshakeToDC.cpp

```diff
@@ -0,0 +1,612 @@
+  Value token;
+  ValueRange data;
+};
+
+// Unpack a !dc.value<...> into a DCTuple.
+static DCTuple unpack(OpBuilder &b, Value v) {
+  if (v.getType().isa<dc::ValueType>())
+    return DCTuple(b.create<dc::UnpackOp>(v.getLoc(), v));
+  else {
+    assert(v.getType().isa<dc::TokenType>() && "Expected a dc::TokenType");
+    return DCTuple(v, ValueRange{});
+  }
+}
+
+static Value pack(OpBuilder &b, Value token, ValueRange data) {
+  if (data.empty())
+    return token;
+  return b.create<dc::PackOp>(token.getLoc(), token, data);
+}
+
+class DCTypeConverter : public TypeConverter {
+public:
+  DCTypeConverter() {
+    addConversion([](Type type) -> Type {
+      if (type.isa<NoneType>())
+        return dc::TokenType::get(type.getContext());
+      else
+        return dc::ValueType::get(type.getContext(), type);
+    });
+    addConversion([](ValueType type) { return type; });
+    addConversion([](TokenType type) { return type; });
+
+    addTargetMaterialization(
+        [&](mlir::OpBuilder &builder, mlir::Type resultType,
+            mlir::ValueRange inputs,
+            mlir::Location loc) -> std::optional<mlir::Value> {
+          if (inputs.size() != 1)
+            return std::nullopt;
+
+          // Materialize !dc.value<> -> !dc.token
+          if (resultType.isa<dc::TokenType>() &&
+              inputs.front().getType().isa<dc::ValueType>())
+            return unpack(builder, inputs.front()).token;
+
+          // Materialize !dc.token -> !dc.value<>
+          auto vt = resultType.dyn_cast<dc::ValueType>();
+          if (vt && vt.getInnerTypes().empty())
+            return pack(builder, inputs.front(), ValueRange{});
+
+          return inputs[0];
+        });
+
+    addSourceMaterialization(
+        [&](mlir::OpBuilder &builder, mlir::Type resultType,
+            mlir::ValueRange inputs,
+            mlir::Location loc) -> std::optional<mlir::Value> {
+          if (inputs.size() != 1)
+            return std::nullopt;
+
+          // Materialize !dc.value<> -> !dc.token
+          if (resultType.isa<dc::TokenType>() &&
+              inputs.front().getType().isa<dc::ValueType>())
+            return unpack(builder, inputs.front()).token;
+
+          // Materialize !dc.token -> !dc.value<>
+          auto vt = resultType.dyn_cast<dc::ValueType>();
+          if (vt && vt.getInnerTypes().empty())
+            return pack(builder, inputs.front(), ValueRange{});
+
+          return inputs[0];
+        });
+  }
+};
+
+template <typename OpTy>
+class DCOpConversionPattern : public OpConversionPattern<OpTy> {
+public:
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  using OpAdaptor = typename OpConversionPattern<OpTy>::OpAdaptor;
+
+  DCOpConversionPattern(MLIRContext *context, TypeConverter &typeConverter,
+                        ConvertedOps *convertedOps)
+      : OpConversionPattern<OpTy>(typeConverter, context),
+        convertedOps(convertedOps) {}
+  mutable ConvertedOps *convertedOps;
```

Comment:
I'm not a fan of `mutable`, but I guess there is no way around it.

Reply:
Yea... i'm super open to other suggestions but this was the easiest way i found to mark operations as legal - either that, or add an attribute to the new ops.

Reply:
It's a shame that one cannot register listeners on the rewriter to get out the set of created ops in a clean manner.